src/librustc_typeck/check/upvar.rs

```diff
@@ -132,14 +132,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         };
 
         self.tcx.with_freevars(closure_node_id, |freevars| {
+            let mut freevar_list: Vec<ty::UpvarId> = Vec::new();
```

Comment:
Nit: I would use `Vec::with_capacity` -- or, perhaps even better, convert to an iterator:

```rust
let upvar_ids: Vec<_> = freevars.iter().map(|freevar| ty::UpvarId {..}).collect();
self.tables
                 .borrow_mut()
                 .upvar_list
                 .insert(closure_def_id, upvar_ids);

for upvar_id in &upvar_ids {
    ... // as before
}
```

Reply:
I agree with the with_capacity part to save on the vector reallocation but by using freevars.iter().map(...).collect() etc. we are iterating over the freevars twice (once here and other time in for freevar in freevars {.. ).  Wouldn't you say that's less performant ?

Reply:
Potentially. It seems unlikely to matter in practice, so I was more concerned with what "read better".

You could of course do both at once by adding a `inspect` in the chain:

```rust
let upvar_ids: Vec<_> = freevars.iter()
    .map(|freevar| ty::UpvarId {..})
    .inspect(|&upvar_id| {
        let capture_kind = /* as before */;
        self.tables.borrow_mut().insert(upvar_id, capture_kind);
    })
    .collect();
```

I tend to prefer not to use explicit `with_capacity` calls and instead use iterators/collector, but either one is fine. Matter of personal preference I suppose.